optuna/integration/wandb.py

```diff
@@ -124,12 +134,51 @@ def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -
                 names = [*self._metric_name]
 
         metrics = {name: value for name, value in zip(names, trial.values)}
+
+        if self._as_multirun:
+            metrics["trial_number"] = trial.number
+
         attributes = {"direction": [d.name for d in study.directions]}
 
-        wandb.config.update(attributes)
-        wandb.log({**trial.params, **metrics}, step=trial.number)
+        step = trial.number if wandb.run else None
+        run = wandb.run
 
-    def _initialize_run(self) -> None:
-        """Initializes Weights & Biases run."""
+        # When an user doesn't use `track_in_wandb` decorator.
+        # Might create an extra run if an user logs in wandb but doesn't use the decorator.
+
+        if not run:
+            run = self._initialize_run()
+            run.name = f"trial/{trial.number}/{run.name}"
 
-        wandb.init(**self._wandb_kwargs)
+        run.log({**trial.params, **metrics}, step=step)
+
+        run.config.update({**attributes, **trial.params})
+
+        if self._as_multirun:
+            run.tags = tuple(self._wandb_kwargs.get("tags", ())) + (study.study_name,)
+            run.finish()
```

Comment:
I think this is happening because  config.update tries to update the values of trial.params for a single run. Apparently that is forbidden by default, since config is meant to be static. In that case, for keeping consistency of logged data between two modes, we should either: remove `trial.params` in both the cases, or use `allow_value_change=True` when `as_multirun=False` while logging `trial.params`. IMO, either cases would work fine. But, just to give the user an extra option on how they should retrieve their parameters, maybe we should log `trial.params` in the config in both cases.

Reply:
If config was meant to be static, let's keep it that way, especially since we are using it only to get param importance plot in W&B, which does not work with `as_multirun=False` anyway. For consistency, we are logging params with `wandb.log` in both modes, so I'm fine with logging them in config conditionally.